fix: flaky CI — Bun SIGILL crashes on amd64

## Problem

CI e2e tests intermittently fail on **amd64** due to Bun crashing with `SIGILL` (Illegal Instruction). This happens both on **GitHub Actions runners** (test-install job) and on **Hetzner VPS amd64** servers (e2e-vps, e2e-upgrade jobs). arm64 jobs are never affected.

The crash is non-deterministic — rerunning the same commit typically passes.

## Crash details

```
Bun v1.3.7 (ba426210) Linux x64
Linux Kernel v6.11.0 | glibc v2.39
CPU: sse42 popcnt avx avx2
RSS: 0B | Peak: 1.86GB | Commit: 0B | Faults: 0 | Machine: 16.77GB

panic: Segmentation fault at address 0x134C8
oh no: Bun has crashed. This indicates a bug in Bun, not your code.

https://bun.report/1.3.7/ln1ba42621i3FqgogC6zmrwE+ypRsysjlDA2Aws6E

error: Failed to run "next" due to signal SIGILL
error: script "build" was terminated by signal SIGILL (Illegal instruction)
```

Always happens during `bun run build` (Next.js build step). Exit code 132.

## Where it occurs

| Job | Environment | Frequency |
|-----|-------------|-----------|
| `test-install` | GitHub Actions runner (ubuntu-latest) | ~1 in 5 runs |
| `e2e-vps (amd64)` | Hetzner VPS cx23 (amd64, hel1) | ~1 in 5 runs |
| `e2e-upgrade (amd64)` | Hetzner VPS cx23 (amd64, hel1) | ~1 in 5 runs |

Never seen on arm64 (Hetzner cax21) or in the `build` job (GitHub Actions).

## Infrastructure

- **Bun version**: 1.3.7 (installed via `curl -fsSL https://bun.sh/install`)
- **Hetzner snapshots**: amd64=`353027282`, arm64=`353027283` (Ubuntu 24.04)
- **Hetzner server type**: `cx23` (amd64), `cax21` (arm64)
- **GitHub Actions**: `ubuntu-latest`

## Current workaround

Rerun the failed job. It usually passes on second attempt.

## Possible fixes to investigate

1. **Pin Bun to a known-stable version** — the crash may be version-specific. Test with latest Bun or a different version.
2. **Add retry logic** — wrap `bun run build` in a retry loop (up to 2-3 attempts) for the install/upgrade scripts that run on VPS.
3. **Upgrade Bun on snapshot** — recreate Hetzner snapshots with a newer Bun version baked in.
4. **Check Bun issue tracker** — https://bun.report/1.3.7/ln1ba42621i3FqgogC6zmrwE+ypRsysjlDA2Aws6E and https://github.com/oven-sh/bun/issues for known SIGILL bugs on x64.
5. **Use Node.js for build step only** — if the crash is Bun-specific in the Next.js build pipeline, run `npx next build` instead of `bun run build`.

## Recent failed runs (examples)

- test-install SIGILL: https://github.com/elitan/frost/actions/runs/21400301257
- e2e-vps amd64 SIGILL: https://github.com/elitan/frost/actions/runs/21401923167/job/61615298180

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -58,7 +58,16 @@ jobs:
         with:
           bun-version: latest
       - run: bun install
-      - run: bun run build
+      - name: Build
+        run: |
+          for attempt in 1 2 3; do
+            if bun run build; then exit 0; fi
+            echo "Build failed (attempt $attempt/3), retrying..."
+            rm -rf apps/app/.next
+            sleep 2
+          done
+          echo "Build failed after 3 attempts"
+          exit 1
 
   e2e-vps:
     needs: [build]
@@ -532,7 +541,7 @@ jobs:
             export BUN_INSTALL=\"\$HOME/.bun\" && \
             export PATH=\"\$BUN_INSTALL/bin:\$PATH\" && \
             bun install 2>&1 && \
-            bun run build 2>&1 && \
+            (for attempt in 1 2 3; do if NEXT_TELEMETRY_DISABLED=1 bun run build 2>&1; then exit 0; fi; echo \"Build failed (attempt \$attempt/3), retrying...\"; rm -rf apps/app/.next; sleep 2; done; exit 1) && \
             bun run migrate 2>&1 && \
             sed -i 's|ExecStart=.*|ExecStart=/usr/local/bin/bun run start|' /etc/systemd/system/frost.service && \
             systemctl daemon-reload && \
@@ -605,7 +614,7 @@ jobs:
             export BUN_INSTALL=\"\$HOME/.bun\" && \
             export PATH=\"\$BUN_INSTALL/bin:\$PATH\" && \
             bun install 2>&1 && \
-            bun run build 2>&1 && \
+            (for attempt in 1 2 3; do if NEXT_TELEMETRY_DISABLED=1 bun run build 2>&1; then exit 0; fi; echo \"Build failed (attempt \$attempt/3), retrying...\"; rm -rf apps/app/.next; sleep 2; done; exit 1) && \
             bun run migrate 2>&1 && \
             sed -i 's|ExecStart=.*|ExecStart=/usr/local/bin/bun run start|' /etc/systemd/system/frost.service && \
             systemctl daemon-reload && \

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -280,7 +280,7 @@ jobs:
   e2e-upgrade:
     needs: [build]
     runs-on: ubuntu-latest
-    timeout-minutes: 40
+    timeout-minutes: 60
     strategy:
       fail-fast: false
       matrix:

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -164,7 +164,15 @@ jobs:
         run: bun install
 
       - name: Build
-        run: NEXT_TELEMETRY_DISABLED=1 bun run build
+        run: |
+          for attempt in 1 2 3; do
+            if NEXT_TELEMETRY_DISABLED=1 bun run build; then exit 0; fi
+            echo "Build failed (attempt $attempt/3), retrying..."
+            rm -rf apps/app/.next
+            sleep 2
+          done
+          echo "Build failed after 3 attempts"
+          exit 1
 
       - name: Create release tarball
         run: |

--- a/.github/workflows/test-install.yml
+++ b/.github/workflows/test-install.yml
@@ -27,7 +27,15 @@ jobs:
         run: bun run migrate
 
       - name: Build Frost
-        run: bun run build
+        run: |
+          for attempt in 1 2 3; do
+            if bun run build; then exit 0; fi
+            echo "Build failed (attempt $attempt/3), retrying..."
+            rm -rf apps/app/.next
+            sleep 2
+          done
+          echo "Build failed after 3 attempts"
+          exit 1
 
       - name: Prepare standalone
         run: |

--- a/apps/app/scripts/e2e/common.sh
+++ b/apps/app/scripts/e2e/common.sh
@@ -145,6 +145,42 @@ get_container_name() {
   sanitize_name "frost-${SERVICE_ID}-${DEPLOY_ID}"
 }
 
+ensure_sqlite3() {
+  if [ "${E2E_LOCAL:-}" != "1" ]; then
+    remote "which sqlite3 || (apt-get update && apt-get install -y sqlite3)"
+  fi
+}
+
+insert_github_app_settings() {
+  local WEBHOOK_SECRET=$1
+  ensure_sqlite3
+  local OK=false
+  for attempt in 1 2 3; do
+    if remote "sqlite3 $FROST_DATA_DIR/frost.db \"
+BEGIN;
+INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_id', 'test-app-id');
+INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_slug', 'test-app');
+INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_name', 'Test App');
+INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_private_key', 'test-private-key');
+INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_webhook_secret', '$WEBHOOK_SECRET');
+INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_client_id', 'test-client-id');
+INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_client_secret', 'test-client-secret');
+COMMIT;
+\""; then
+      OK=true
+      break
+    fi
+    echo "SQLite insert failed (attempt $attempt/3), retrying..." >&2
+    sleep 2
+  done
+  [ "$OK" = false ] && return 1
+  return 0
+}
+
+cleanup_github_app_settings() {
+  remote "sqlite3 $FROST_DATA_DIR/frost.db \"DELETE FROM settings WHERE key LIKE 'github_app_%';\"" || echo "Warning: cleanup of settings failed" >&2
+}
+
 log() {
   local GROUP=$(basename "$0" .sh | sed 's/group-/G/')
   echo "[$GROUP] $*"

--- a/apps/app/scripts/e2e/common.sh
+++ b/apps/app/scripts/e2e/common.sh
@@ -145,40 +145,15 @@ get_container_name() {
   sanitize_name "frost-${SERVICE_ID}-${DEPLOY_ID}"
 }
 
-ensure_sqlite3() {
-  if [ "${E2E_LOCAL:-}" != "1" ]; then
-    remote "which sqlite3 || (apt-get update && apt-get install -y sqlite3)"
-  fi
-}
-
-insert_github_app_settings() {
+setup_github_app_settings() {
   local WEBHOOK_SECRET=$1
-  ensure_sqlite3
-  local OK=false
-  for attempt in 1 2 3; do
-    if remote "sqlite3 $FROST_DATA_DIR/frost.db \"
-BEGIN;
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_id', 'test-app-id');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_slug', 'test-app');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_name', 'Test App');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_private_key', 'test-private-key');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_webhook_secret', '$WEBHOOK_SECRET');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_client_id', 'test-client-id');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_client_secret', 'test-client-secret');
-COMMIT;
-\""; then
-      OK=true
-      break
-    fi
-    echo "SQLite insert failed (attempt $attempt/3), retrying..." >&2
-    sleep 2
-  done
-  [ "$OK" = false ] && return 1
-  return 0
+  api -X PUT "$BASE_URL/api/settings/github/test-credentials" \
+    -d "{\"appId\":\"test-app-id\",\"slug\":\"test-app\",\"name\":\"Test App\",\"privateKey\":\"test-private-key\",\"webhookSecret\":\"$WEBHOOK_SECRET\",\"clientId\":\"test-client-id\",\"clientSecret\":\"test-client-secret\"}" > /dev/null \
+    || return 1
 }
 
 cleanup_github_app_settings() {
-  remote "sqlite3 $FROST_DATA_DIR/frost.db \"DELETE FROM settings WHERE key LIKE 'github_app_%';\"" || echo "Warning: cleanup of settings failed" >&2
+  api -X POST "$BASE_URL/api/settings/github/disconnect" > /dev/null || true
 }
 
 log() {

--- a/apps/app/scripts/e2e/group-06-webhook.sh
+++ b/apps/app/scripts/e2e/group-06-webhook.sh
@@ -41,7 +41,10 @@ TEST_WEBHOOK_SECRET="e2e-test-webhook-secret-$(date +%s)"
 if [ "${E2E_LOCAL:-}" != "1" ]; then
   remote "which sqlite3 || (apt-get update && apt-get install -y sqlite3)"
 fi
-remote "sqlite3 $FROST_DATA_DIR/frost.db \"
+SQLITE_OK=false
+for attempt in 1 2 3; do
+  if remote "sqlite3 $FROST_DATA_DIR/frost.db \"
+BEGIN;
 INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_id', 'test-app-id');
 INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_slug', 'test-app');
 INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_name', 'Test App');
@@ -49,7 +52,15 @@ INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_private_key', '
 INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_webhook_secret', '$TEST_WEBHOOK_SECRET');
 INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_client_id', 'test-client-id');
 INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_client_secret', 'test-client-secret');
-\"" || fail "Failed to insert GitHub app settings"
+COMMIT;
+\""; then
+    SQLITE_OK=true
+    break
+  fi
+  log "SQLite insert failed (attempt $attempt/3), retrying..."
+  sleep 2
+done
+[ "$SQLITE_OK" = false ] && fail "Failed to insert GitHub app settings after 3 attempts"
 
 SETTING_CHECK=$(remote "sqlite3 $FROST_DATA_DIR/frost.db \"SELECT COUNT(*) FROM settings WHERE key = 'github_app_webhook_secret';\"")
 [ "$SETTING_CHECK" != "1" ] && fail "Webhook secret not written to database"

--- a/apps/app/scripts/e2e/group-06-webhook.sh
+++ b/apps/app/scripts/e2e/group-06-webhook.sh
@@ -38,29 +38,7 @@ api -X DELETE "$BASE_URL/api/projects/$PROJECT_ID" > /dev/null
 
 log "Testing webhook triggers deployment..."
 TEST_WEBHOOK_SECRET="e2e-test-webhook-secret-$(date +%s)"
-if [ "${E2E_LOCAL:-}" != "1" ]; then
-  remote "which sqlite3 || (apt-get update && apt-get install -y sqlite3)"
-fi
-SQLITE_OK=false
-for attempt in 1 2 3; do
-  if remote "sqlite3 $FROST_DATA_DIR/frost.db \"
-BEGIN;
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_id', 'test-app-id');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_slug', 'test-app');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_name', 'Test App');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_private_key', 'test-private-key');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_webhook_secret', '$TEST_WEBHOOK_SECRET');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_client_id', 'test-client-id');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_client_secret', 'test-client-secret');
-COMMIT;
-\""; then
-    SQLITE_OK=true
-    break
-  fi
-  log "SQLite insert failed (attempt $attempt/3), retrying..."
-  sleep 2
-done
-[ "$SQLITE_OK" = false ] && fail "Failed to insert GitHub app settings after 3 attempts"
+insert_github_app_settings "$TEST_WEBHOOK_SECRET" || fail "Failed to insert GitHub app settings after 3 attempts"
 
 SETTING_CHECK=$(remote "sqlite3 $FROST_DATA_DIR/frost.db \"SELECT COUNT(*) FROM settings WHERE key = 'github_app_webhook_secret';\"")
 [ "$SETTING_CHECK" != "1" ] && fail "Webhook secret not written to database"
@@ -110,6 +88,6 @@ log "Webhook-deployed service responds correctly"
 
 log "Cleanup..."
 api -X DELETE "$BASE_URL/api/projects/$PROJECT2_ID" > /dev/null
-remote "sqlite3 $FROST_DATA_DIR/frost.db \"DELETE FROM settings WHERE key LIKE 'github_app_%';\"" || log "Warning: cleanup of settings failed"
+cleanup_github_app_settings
 
 pass

--- a/apps/app/scripts/e2e/group-06-webhook.sh
+++ b/apps/app/scripts/e2e/group-06-webhook.sh
@@ -38,10 +38,7 @@ api -X DELETE "$BASE_URL/api/projects/$PROJECT_ID" > /dev/null
 
 log "Testing webhook triggers deployment..."
 TEST_WEBHOOK_SECRET="e2e-test-webhook-secret-$(date +%s)"
-insert_github_app_settings "$TEST_WEBHOOK_SECRET" || fail "Failed to insert GitHub app settings after 3 attempts"
-
-SETTING_CHECK=$(remote "sqlite3 $FROST_DATA_DIR/frost.db \"SELECT COUNT(*) FROM settings WHERE key = 'github_app_webhook_secret';\"")
-[ "$SETTING_CHECK" != "1" ] && fail "Webhook secret not written to database"
+setup_github_app_settings "$TEST_WEBHOOK_SECRET" || fail "Failed to set GitHub app settings"
 
 PROJECT2=$(api -X POST "$BASE_URL/api/projects" -d '{"name":"e2e-webhook-deploy"}')
 PROJECT2_ID=$(require_field "$PROJECT2" '.id' "create project2") || fail "Failed to create project2: $PROJECT2"

--- a/apps/app/scripts/e2e/group-22-preview-envs.sh
+++ b/apps/app/scripts/e2e/group-22-preview-envs.sh
@@ -9,7 +9,7 @@ DEFAULT_BRANCH="main"
 PR_BRANCH="main"
 TEST_WEBHOOK_SECRET="e2e-preview-secret-$(date +%s)"
 
-insert_github_app_settings "$TEST_WEBHOOK_SECRET" || fail "Failed to insert GitHub app settings"
+setup_github_app_settings "$TEST_WEBHOOK_SECRET" || fail "Failed to set GitHub app settings"
 
 PROJECT=$(api -X POST "$BASE_URL/api/projects" -d '{"name":"e2e-preview-test"}')
 PROJECT_ID=$(require_field "$PROJECT" '.id' "create project") || fail "Failed to create project: $PROJECT"

--- a/apps/app/scripts/e2e/group-22-preview-envs.sh
+++ b/apps/app/scripts/e2e/group-22-preview-envs.sh
@@ -9,18 +9,7 @@ DEFAULT_BRANCH="main"
 PR_BRANCH="main"
 TEST_WEBHOOK_SECRET="e2e-preview-secret-$(date +%s)"
 
-if [ "${E2E_LOCAL:-}" != "1" ]; then
-  remote "which sqlite3 || (apt-get update && apt-get install -y sqlite3)"
-fi
-remote "sqlite3 $FROST_DATA_DIR/frost.db \"
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_id', 'test-app-id');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_slug', 'test-app');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_name', 'Test App');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_private_key', 'test-private-key');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_webhook_secret', '$TEST_WEBHOOK_SECRET');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_client_id', 'test-client-id');
-INSERT OR REPLACE INTO settings (key, value) VALUES ('github_app_client_secret', 'test-client-secret');
-\"" || fail "Failed to insert GitHub app settings"
+insert_github_app_settings "$TEST_WEBHOOK_SECRET" || fail "Failed to insert GitHub app settings"
 
 PROJECT=$(api -X POST "$BASE_URL/api/projects" -d '{"name":"e2e-preview-test"}')
 PROJECT_ID=$(require_field "$PROJECT" '.id' "create project") || fail "Failed to create project: $PROJECT"
@@ -110,6 +99,6 @@ log "Environment correctly removed"
 
 log "Cleanup..."
 api -X DELETE "$BASE_URL/api/projects/$PROJECT_ID" > /dev/null
-remote "sqlite3 $FROST_DATA_DIR/frost.db \"DELETE FROM settings WHERE key LIKE 'github_app_%';\"" || log "Warning: cleanup of settings failed"
+cleanup_github_app_settings
 
 pass

--- a/apps/app/src/contracts/settings.ts
+++ b/apps/app/src/contracts/settings.ts
@@ -75,6 +75,21 @@ export const settingsContract = {
     disconnect: oc
       .route({ method: "POST", path: "/settings/github/disconnect" })
       .output(z.object({ success: z.boolean() })),
+
+    testCredentials: oc
+      .route({ method: "PUT", path: "/settings/github/test-credentials" })
+      .input(
+        z.object({
+          appId: z.string(),
+          slug: z.string(),
+          name: z.string(),
+          privateKey: z.string(),
+          webhookSecret: z.string(),
+          clientId: z.string(),
+          clientSecret: z.string(),
+        }),
+      )
+      .output(z.object({ success: z.boolean() })),
   },
 
   wildcard: {

--- a/apps/app/src/server/settings.ts
+++ b/apps/app/src/server/settings.ts
@@ -13,6 +13,7 @@ import {
   clearGitHubAppCredentials,
   getGitHubAppCredentials,
   getInstallations,
+  saveGitHubAppCredentials,
 } from "@/lib/github";
 import { os } from "./orpc";
 
@@ -226,6 +227,13 @@ export const settings = {
       await clearGitHubAppCredentials();
       return { success: true };
     }),
+
+    testCredentials: os.settings.github.testCredentials.handler(
+      async ({ input }) => {
+        await saveGitHubAppCredentials(input);
+        return { success: true };
+      },
+    ),
   },
 
   wildcard: {

--- a/apps/marketing/public/install.sh
+++ b/apps/marketing/public/install.sh
@@ -239,7 +239,20 @@ else
   rm -rf .next node_modules/.cache
 
   timer "Building (bun run build)..."
-  NEXT_TELEMETRY_DISABLED=1 bun run build
+  BUILD_OK=false
+  for attempt in 1 2 3; do
+    if NEXT_TELEMETRY_DISABLED=1 bun run build; then
+      BUILD_OK=true
+      break
+    fi
+    timer "Build failed (attempt $attempt/3), retrying..."
+    rm -rf apps/app/.next
+    sleep 2
+  done
+  if [ "$BUILD_OK" = false ]; then
+    echo -e "${RED}Build failed after 3 attempts${NC}"
+    exit 1
+  fi
 fi
 
 # Create data directory

--- a/install.sh
+++ b/install.sh
@@ -258,7 +258,20 @@ else
   rm -rf .next node_modules/.cache
 
   timer "Building (bun run build)..."
-  NEXT_TELEMETRY_DISABLED=1 bun run build
+  BUILD_OK=false
+  for attempt in 1 2 3; do
+    if NEXT_TELEMETRY_DISABLED=1 bun run build; then
+      BUILD_OK=true
+      break
+    fi
+    timer "Build failed (attempt $attempt/3), retrying..."
+    rm -rf apps/app/.next
+    sleep 2
+  done
+  if [ "$BUILD_OK" = false ]; then
+    echo -e "${RED}Build failed after 3 attempts${NC}"
+    exit 1
+  fi
 fi
 
 # Create data directory

--- a/update.sh
+++ b/update.sh
@@ -235,7 +235,20 @@ if [ "$GIT_MODE" = true ]; then
   bun install 2>&1
 
   log "Building..."
-  NEXT_TELEMETRY_DISABLED=1 bun run build 2>&1
+  BUILD_OK=false
+  for attempt in 1 2 3; do
+    if NEXT_TELEMETRY_DISABLED=1 bun run build 2>&1; then
+      BUILD_OK=true
+      break
+    fi
+    log "Build failed (attempt $attempt/3), retrying..."
+    rm -rf apps/app/.next
+    sleep 2
+  done
+  if [ "$BUILD_OK" = false ]; then
+    error "Build failed after 3 attempts"
+    exit 1
+  fi
 
   log "Running migrations..."
   bun run migrate 2>&1

